### PR TITLE
Add options for configuring completion annotations

### PR DESCRIPTION
**Pull request by cichli**
This is looking good but there's one issue - cljs-tooling returns a value for `:ns` when the candidate is itself a namespace whereas compliment returns `nil`:

 

I think that `:ns` should be returned if the candidate is a namespace alias but elided if the candidate is an actual namespace. Will fix this in the cljs-tooling PR and open a compliment issue (it returns a type of `class` with no value for namespace atm).

I'll add something to the README in the autocompletion section too - let me know if you're happy with it first.

### CODE REVIEW

**Comment by bbatsov** [comment]
> I think that :ns should be returned if the candidate is a namespace alias but elided if the candidate is an actual namespace. Will fix this in the cljs-tooling PR and open a compliment issue (it returns a type of class with no value for namespace atm).

I totally agree.

File: cider-interaction.el [hunk 1]
```diff
@@ -125,6 +125,51 @@ which will use the default REPL connection."
   :group 'cider
   :package-version '(cider . "0.8.0"))
 
+(defcustom cider-annotate-completion-function
+  '(lambda (type ns)
```
Comment (bbatsov) [inline cider-interaction.el]: I'm not fond of using `lambda`s in defcustoms. We can have a default function which will be referred here.

File: cider-interaction.el [hunk 2]
```diff
@@ -799,37 +844,45 @@ form, with symbol at point replaced by __prefix__."
...
 (defun cider-complete (str)
   "Complete STR with context at point."
   (let* ((context (cider-completion-get-context))
          (candidates (cider-sync-request:complete str context)))
-    (map 'list #'cider-completion--parse-candidate-map candidates)))
+    (mapcar #'cider-completion--parse-candidate-map candidates)))
+
+(defun cider-completion--get-candidate-type (symbol)
+  (let ((type (get-text-property 0 'type symbol)))
+    (or (second (assoc type cider-completion-annotations-alist))
```
Comment (bbatsov) [inline cider-interaction.el]: `second` -> `cl-second`

File: cider-util.el [hunk 3]
```diff
@@ -167,6 +167,10 @@ objects."
               (cons el el)))
           candidates))
 
+(defun cider-namespace-qualified-p (sym)
+  "Return t if sym is namespace-qualified."
```
Comment (bbatsov) [inline cider-util.el]: sym -> SYM

File: cider-interaction.el [hunk 4]
```diff
@@ -125,6 +125,51 @@ which will use the default REPL connection."
...
+  "Controls how the annotations for completion candidates are formatted.
+
+Must be a function that takes two arguments: the abbreviation of the
+candidate type according to `cider-completion-annotations-alist' and the
+candidate's namespace."
+  :type 'function
+  :group 'cider
+  :package-version '(cider . "0.9.0"))
+
+(defcustom cider-completion-annotations-alist '(("class" "c")
```
Comment (bbatsov) [inline cider-interaction.el]: Guess you can place the list of the second line. This makes the code a bit more readable IMO.

**Comment by bbatsov** [comment]
Apart from my minor remarks, everything looks good. You can go ahead and update the README.

**Comment by cichli** [comment]
Done :+1:

File: README.md [hunk 5]
```diff
@@ -492,6 +492,13 @@ inside of source code and REPL buffers. This can be done globally, like so --
 When `company-mode` is thus enabled, it will receive completion information
 from `cider-complete-at-point`, and requires no additional setup or plugins.
 
+Completion candidates will be annotated by default when using company-mode, with
```
Comment (bbatsov) [inline README.md]: Btw, the annotations will be visible in the standard Emacs completions as well. You can also add some screenshot here.

**Comment by cichli** [comment]
Should be good now.

**Comment by bbatsov** [comment]
:+1:

**Pull request merged**